clang-tools-extra/clang-tidy/bugprone/ComparePointerToMemberVirtualFunctionCheck.cpp

```diff
@@ -0,0 +1,89 @@
+//===--- ComparePointerToMemberVirtualFunctionCheck.cpp - clang-tidy ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "ComparePointerToMemberVirtualFunctionCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/ASTTypeTraits.h"
+#include "clang/AST/DeclCXX.h"
+#include "clang/AST/OperationKinds.h"
+#include "clang/AST/Type.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "clang/ASTMatchers/ASTMatchersMacros.h"
+#include "clang/Basic/Diagnostic.h"
+#include "clang/Basic/DiagnosticIDs.h"
+#include "llvm/ADT/SmallVector.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+AST_MATCHER(CXXMethodDecl, isVirtual) { return Node.isVirtual(); }
+
+void ComparePointerToMemberVirtualFunctionCheck::registerMatchers(
+    MatchFinder *Finder) {
+
+  auto DirectMemberPointer = unaryOperator(
+      allOf(hasOperatorName("&"),
+            hasUnaryOperand(declRefExpr(to(cxxMethodDecl(isVirtual()))))));
+  auto IndirectMemberPointer = ignoringImpCasts(declRefExpr());
+
+  auto BinaryOperatorMatcher = [](auto &&Matcher) {
```

Comment:
Avoid matching same thing twice, this can be writeln as:
```
binaryOperator(hasAnyOperatorName("==", "!="),
               hasEitherOperand(hasType(memberPointerType(pointee(functionType())))),
               anyOf(binaryOperator(hasEitherOperand(DirectMemberPointer)).bind("direct_member_pointer"),
                     binaryOperator(hasEitherOperand(IndirectMemberPointer)).bind("indirect_member_pointer")),
               unless(hasEitherOperand(castExpr(hasCastKind(CK_NullToMemberPointer))))
              )
```

Reply:
binaryOperator duplicate in this matcher. Is there any way to get parent node in later check? Then I can bind to operand and get operator location?

Reply:
You could split this, for example bind DirectMemberPointer/IndirectMemberPointer expressions and use them for logic control, and bind binarOperator once.

```
binaryOperator(hasAnyOperatorName("==", "!="),
               hasEitherOperand(hasType(memberPointerType(pointee(functionType())))),
               hasEitherOperand(expr(anyOf(DirectMemberPointer.bind("direct_member_pointer"),
                                                                IndirectMemberPointer.bind("indirect_member_pointer")))),
               unless(hasEitherOperand(castExpr(hasCastKind(CK_NullToMemberPointer))))
              ).bind("bo")
```